sam_view.c

```diff
@@ -708,36 +799,39 @@ static readpart which_readpart(const bam1_t *b)
     }
 }
 
-// Transform a bam1_t record into a string with the FASTQ representation of it
-// @returns false for error, true for success
-static bool bam1_to_fq(const bam1_t *b, kstring_t *linebuf, const bam2fq_state_t *state)
+/*
+ * parse the length part from the index-format string
+ */
+static int getLength(char **s)
 {
-    int i;
-    int32_t qlen = b->core.l_qseq;
-    assert(qlen >= 0);
-    uint8_t *seq;
-    uint8_t *qual = bam_get_qual(b);
-    const uint8_t *oq = NULL;
-    if (state->use_oq) {
-        oq = bam_aux_get(b, "OQ");
-        if (oq) oq++; // skip tag type
+    int n = 0;
+    while (**s) {
+        if (**s == '*') { n=-1; (*s)++; break; }
+        if ( !isdigit(**s)) break;
+        if (isdigit(**s)) { n = n*10 + ((**s)-'0'); }
+        (*s)++;
     }
-    bool has_qual = (qual[0] != 0xff || (state->use_oq && oq)); // test if there is quality
+    return n;
+}
+
+static bool make_fq_line(const bam1_t *rec, char *seq, char *qual, kstring_t *linebuf, const bam2fq_state_t *state)
+{
+    int i;
 
     linebuf->l = 0;
     // Write read name
-    readpart readpart = which_readpart(b);
     kputc(state->filetype == FASTA? '>' : '@', linebuf);
-    kputs(bam_get_qname(b), linebuf);
+    kputs(bam_get_qname(rec), linebuf);
     // Add the /1 /2 if requested
     if (state->has12) {
+        readpart readpart = which_readpart(rec);
         if (readpart == READ_1) kputs("/1", linebuf);
         else if (readpart == READ_2) kputs("/2", linebuf);
     }
     if (state->copy_tags) {
         for (i = 0; copied_tags[i]; ++i) {
             uint8_t *s;
-            if ((s = bam_aux_get(b, copied_tags[i])) != 0) {
+            if ((s = bam_aux_get(rec, copied_tags[i])) != 0) {
```

Comment:
There is an implicit assumption here that all copied_tags[] array entries are for :Z: tags.  While this is true in the standard, it may not be true in actual data files.  This could lead to calling kputs on a non-string and running off the end of the bam record causing a crash or leaking the contents of the remainder of a malloced block of data.

Please add a check for ```*s == 'Z'```.